misc/multistream-select/src/protocol.rs

```diff
@@ -67,11 +68,40 @@ impl From<Version> for HeaderLine {
 }
 
 /// A protocol (name) exchanged during protocol negotiation.
-#[derive(Clone, Debug, PartialEq, Eq)]
-pub struct Protocol(Bytes);
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+pub struct Protocol(Cow<'static, str>); // TODO: Instead of `Cow`, we should probably be storing `Arc<str>`
```

Comment:
How about we start with as is and introduce an `Arc<str>` if necessary? None of it is exposed, thus easy to change in a non-breaking way.

Reply:
I am not fully sure. It improves performance for cases where we dynamically construct a `Protocol`. Ideally, users use `from_static` as much as possible so it is kind of a niche usecase but I still want it to be performant.

It is not much work so I think I am just gonna do it. I don't see why a protocol string should ever be mutable which is the only thing that becomes impossible with `Arc<str>`.

Reply:
An `Arc<str>` would not be enough, right? `Arc` would require a heap allocation, no? So you would have `Either<&'static str, Arc<String>>` instead?

Reply:
Yes, we would have an `Either` of `&'static str` and `Arc<String`>.